src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets

```diff
@@ -196,9 +196,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <!-- The defaults currently root non-framework assemblies, which
          is a no-op for portable apps. If we later support more ways
          to customize the behavior we can allow linking portable apps
-         in some cases. If we're not running ILLink because e.g. this
-         is a NativeAOT app, value of SelfContained doesn't matter. -->
-    <NETSdkError Condition="'$(RunILLink)' != 'false' And '$(SelfContained)' != 'true'" ResourceName="ILLinkNotSupportedError" />
+         in some cases. -->
+    <NETSdkError Condition="'$(SelfContained)' != 'true'" ResourceName="ILLinkNotSupportedError" />
```

Comment:
This is a breaking change, right? A command that didn't fail before will now.

Reply:
> This can lead to a bad failure mode because SelfContained is the thing that ensure ILC gets all the references to assemblies it needs.

Can we instead base this off of `SelfContained` OR `PublishSelfContained`? If either of those are set, ensure ILC gets all the references to assemblies it needs. 

Or just completely deprecate `SelfContained` since there is so much confusion and broken things here. See also


https://github.com/dotnet/sdk/issues/32272
https://github.com/dotnet/sdk/issues/32277

Reply:
> I really don't care what the fix is

Why not take my suggested fix then?

Check **both** `SelfContained` and `PublishSelfContained` to  ensure ILC gets all the references to assemblies it needs.

Reply:
You mean update all places in the sdk that check self contained to also check publishselfcontained? I don't even know why we have two in the first place...

Reply:
To be clear the problem is that the assembly list that comes from the sdk doesn't have the necessary assemblies.

Reply:
> You mean update all places in the sdk that check self contained to also check publishselfcontained? I don't even know why we have two in the first place...

Read those 2 issues. I don't know why either.

Reply:
> Read those 2 issues. I don't know why either.

It looks like a hard fix that someone who understands the sdk will need to fix. This aligns aot with trimmed and we're no worse off. The error cmessage ould be improved but it's correct in principle.

Reply:
> and we're no worse off

That's not true for someone who is using the current behavior of `dotnet msbuild /t:Publish`.

> It looks like a hard fix that someone who understands the sdk will need to fix.

Why? I don't understand what is hard about the fix. In the place that says `if (SelfContained) { AddAllAssemblies(); }`, change it to be `if (SelfContained || PublishSelfContained) { AddAllAssemblies(); }`. The problem on those issues is that the conversation escalated to the larger "why do we have both?" conversation, which never got satisified answers.

Reply:
> I think this change moves us in the right direction.

I understand that this change makes PublishAot consistent with PublishTrimmed. But it makes it consistent with the wrong behavior. It is pointless to tell the user "Did you set SelfContained?" when SelfContained should be the default when setting PublishAot. The UX is bad here. If you want to be consistent, we should also fail if the user doesn't manually set the RID. And if they don't manually set PublishTrimmed and PublishSingleFile when setting PublishAot.

As a user, I should just need to set a single property: `PublishAot`, and all these other things (SelfContained, RID, PublishTrimmed, PublishSingleFile, etc) are all defaulted to a value that works. I shouldn't have to set 2 properties. We had this conversation last year and came to that agreement. From @nagilson's emailed meeting notes of that meeting:

> Here is the conclusion from the meeting.
For PublishReadyToRun, the default will not be SelfContained. This is a breaking change and will be conditioned on the new .NET 8 TFM.
**For PublishSingleFile, PublishTrimmed, and PublishAot: PublishSelfContained will be the default. There will be no warning saying you should set SelfContained explicitly.** 

That meeting was motivated by this even larger discussion: https://github.com/dotnet/sdk/issues/30104.